tests/ProjectJoinRequestsMfaValidationTest.php

```diff
@@ -0,0 +1,172 @@
+            $this->normalUserClient->deleteMyProjectJoinRequest($joinRequest['id']);
+        }
+
+        foreach ($this->client->listMyProjectJoinRequests() as $joinRequest) {
+            $this->client->deleteMyProjectJoinRequest($joinRequest['id']);
+        }
+    }
+
+    public function testSuperAdminWithoutMfaCannotRequestAccess()
+    {
+        $projectId = $this->createProjectWithAdminHavingMfaEnabled();
+
+        $this->normalUserWithMfaClient->updateOrganization(
+            $this->organization['id'],
+            [
+                'mfaRequired' => 1,
+                'allowAutoJoin' => 0,
+            ]
+        );
+
+        $joinRequests = $this->normalUserWithMfaClient->listProjectJoinRequests($projectId);
+        $this->assertCount(0, $joinRequests);
+
+        try {
+            $this->client->requestAccessToProject($projectId);
+            $this->fail('Requesting access to a project should produce error');
+        } catch (ClientException $e) {
+            $this->assertEquals(400, $e->getCode());
+            $this->assertContains('Project requires users to have multi-factor authentication enabled', $e->getMessage());
+        }
+
+        $joinRequests = $this->normalUserWithMfaClient->listProjectJoinRequests($projectId);
+        $this->assertCount(0, $joinRequests);
+    }
+
+    public function testJoinRequestOfUserWithoutMfaCannotBeApproved()
+    {
+        $projectId = $this->createProjectWithAdminHavingMfaEnabled();
+
+        $this->normalUserWithMfaClient->updateOrganization(
+            $this->organization['id'],
+            [
+                'allowAutoJoin' => 0,
+            ]
+        );
+
+        $joinRequests = $this->client->listMyProjectJoinRequests();
+        $this->assertCount(0, $joinRequests);
+
+        $joinRequest = $this->client->requestAccessToProject($projectId);
+
+        $joinRequests = $this->client->listMyProjectJoinRequests();
+        $this->assertCount(1, $joinRequests);
+
+        $this->assertEquals($joinRequest, reset($joinRequests));
+
+        $this->normalUserWithMfaClient->updateOrganization(
+            $this->organization['id'],
+            [
+                'mfaRequired' => 1,
+            ]
+        );
+
+        try {
+            $this->normalUserWithMfaClient->approveProjectJoinRequest($projectId, $joinRequest['id']);
+            $this->fail('Approving a join request should produce error');
+        } catch (ClientException $e) {
+            $this->assertEquals(400, $e->getCode());
+            $this->assertContains('Project requires users to have multi-factor authentication enabled', $e->getMessage());
+        }
+
+        $joinRequests = $this->client->listMyProjectJoinRequests();
+        $this->assertCount(1, $joinRequests);
+
+        $projectUser = $this->findProjectUser($projectId, $this->superAdmin['email']);
+        $this->assertNull($projectUser);
+    }
+
+    public function testJoinRequestReject()
```

Comment:
šlo by líp pojmenovat jako `joinRequestThatCannotBeApprovedBacauseOfMfaCanBeRejected`

Reply:
(už jsem to viděl i někde nahoře ještě jednou)